tensorpipe/test/channel/channel_test.h

```diff
@@ -20,71 +20,168 @@
 #include <tensorpipe/test/peer_group.h>
 #include <tensorpipe/transport/uv/context.h>
 
-class DataWrapper {
+template <typename TBuffer>
+class DataWrapper {};
+
+template <>
+class DataWrapper<tensorpipe::CpuBuffer> {
  public:
-  virtual ~DataWrapper() = default;
-  virtual void* data() = 0;
-  virtual size_t size() = 0;
-  virtual void wrap(const void* ptr) = 0;
-  virtual void unwrap(void* ptr) = 0;
+  explicit DataWrapper(size_t length) : vector_(length) {}
+
+  explicit DataWrapper(const std::vector<uint8_t>& v) : vector_(v) {}
+
+  tensorpipe::CpuBuffer buffer() const {
+    return tensorpipe::CpuBuffer{
+        .ptr = const_cast<uint8_t*>(vector_.data()),
+        .length = vector_.size(),
+    };
+  }
+
+  std::vector<uint8_t> unwrap() {
+    return vector_;
+  }
+
+ private:
+  std::vector<uint8_t> vector_;
 };
 
-class IdWrapper : public DataWrapper {
+#if TENSORPIPE_SUPPORTS_CUDA
+
+template <>
+class DataWrapper<tensorpipe::CudaBuffer> {
  public:
-  explicit IdWrapper(size_t len) : size_(len) {}
+  // Non-copyable.
+  DataWrapper(const DataWrapper&) = delete;
+  DataWrapper& operator=(const DataWrapper&) = delete;
 
-  void* data() override {
-    return data_;
+  explicit DataWrapper(size_t length) : length_(length) {
+    if (length_ > 0) {
+      EXPECT_EQ(cudaSuccess, cudaSetDevice(0));
+      EXPECT_EQ(cudaSuccess, cudaStreamCreate(&stream_));
+      EXPECT_EQ(cudaSuccess, cudaMalloc(&cudaPtr_, length_));
+    }
   }
 
-  size_t size() override {
-    return size_;
+  explicit DataWrapper(const std::vector<uint8_t>& v) : DataWrapper(v.size()) {
+    if (length_ > 0) {
+      EXPECT_EQ(
+          cudaSuccess,
+          cudaMemcpy(cudaPtr_, v.data(), length_, cudaMemcpyDefault));
+    }
   }
 
-  void wrap(const void* ptr) override {
-    data_ = const_cast<void*>(ptr);
+  explicit DataWrapper(DataWrapper&& other)
+      : cudaPtr_(other.cudaPtr_),
+        length_(other.length_),
+        stream_(other.stream_) {
+    other.cudaPtr_ = nullptr;
+    other.length_ = 0;
   }
```

Comment:
Once you destroy the stream in the destructor you may want to reset it to `cudaStreamDefault` here.